### PR TITLE
555 mr model interaction events

### PR DESCRIPTION
**Pull request by michaelthatsit**
## Linking

Options:

- fixes: https://github.com/Volumetrics-io/mrjs/issues/555
- fixes: https://github.com/Volumetrics-io/mrjs/issues/529
- fixes: https://github.com/Volumetrics-io/mrjs/issues/526
- fixes: https://github.com/Volumetrics-io/mrjs/issues/399
- fixes: https://github.com/Volumetrics-io/mrjs/issues/400
- Direct touch collision miscalculation. 

## Problem

- models were loading incorrectly. Occasionally generating duplicates, non-deterministically failing to create colliders.
- events weren't being triggered
- visibility was not being set programmatically
- direct touch was broken
- direct touch and gaze/pinch would run at the same time, competing with one another
- weird error message that we'd get on load

## Solution

- made `connected` asynchronous. This prevents `new-entity` and other events from triggering before the entity (such as model) has fully loaded. this also significantly improves load time and prevents duplicate models from being generated
- refactored `ControlSystem` to fire events at the correct time, and disable gaze pinch when the users is directly touching an element.
- 

### Breaking Change

- with async entity loading, @takahirox's getClientBoundingBox performance hack breaks UI responsiveness. it's unclear why, so I've disabled it until we can find a fix. This will temporarily degrade framerate
- `mouseleave` -> `mouseout`. `mouseenter` and `mouseleave` will be enabled in a future PR.

## Notes

*Notes and any associated research or links*

------------

## Required to Merge

- [x] **PASS** - all necessary actions must pass (excluding the auto-skipped ones)
- [x] **TEST IN HEADSET** - [main dev-testing-example](https://github.com/Volumetrics-io/mrjs/tree/main/samples/index.html) and any of the other [examples](https://github.com/Volumetrics-io/mrjs/tree/main/samples/examples) still work as expected
- [ ] **VIDEO** - if this pr changes something visually - post a video here of it in headset-MR and/or on desktop (depending on what it affects) for the reviewer to reference.
- [x] **TITLE** - make sure the pr's title is updated appropriately as it will be used to name the commit on merge
- [ ] **BREAKING CHANGE**
  - **DOCUMENTATION**: This includes any changes to html tags and their components
    - make a pr in the [documentation repo](https://github.com/Volumetrics-io/documentation) that updates the manual docs to match the breaking change
    - link the pr of the documentation repo here: *#pr*
    - that pr must be approved by `@lobau`
  - **SAMPLES/INDEX.HTML**: This includes any changes (html tags or otherwise) that must be done to our landing page submodule as an effect of this pr's updates
    - make a pr in the [mrjs landing page repo](https://github.com/Volumetrics-io/mrjs-landing) that updates the landing page to match the breaking change
    - link the pr of the landing page repo here: *#pr*
    - that pr must be approved by `@hanbollar`

### CODE REVIEW

File: src/core/MRApp.js [hunk 1]
```diff
@@ -33,6 +33,8 @@ window.mobileCheck = function () {
     return mrjsUtils.display.mobileCheckFunction();
 };
 
+const GLOBAL_UPDATE_EVENTS = ['enterxr', 'exitxr', 'load', 'anchored', 'panelupdate', 'engine-started', 'resize'];
```
Comment (hanbollar) [inline src/core/MRApp.js:36]: add documentation note for why here and why order

File: src/core/MREntity.js [hunk 2]
```diff
@@ -393,22 +382,32 @@ export class MREntity extends MRElement {
...
+
+        if (mrjsUtils.physics.initialized) {
+            await this.connected();
+            this.dispatchEvent(new CustomEvent('new-entity', { bubbles: true }));
+            this.loadAttributes();
+        } else {
+            document.addEventListener('engine-started', async (event) => {
+                await this.connected();
+                this.loadAttributes();
+                this.dispatchEvent(new CustomEvent('new-entity', { bubbles: true }));
```
Comment (hanbollar) [inline src/core/MREntity.js:408]: add small comment here explaining why

File: src/core/MRSystem.js [hunk 3]
```diff
@@ -111,6 +97,17 @@ export class MRSystem {
      */
     eventUpdate() {}
 
+    /**
+     *
+     * @param entity
+     */
```
Comment (hanbollar) [inline src/core/MRSystem.js:103]: look at the jsdoc warnings in the files section of github - make sure theyre filled out

File: src/core/MRSystem.js [hunk 4]
```diff
@@ -120,10 +117,10 @@ export class MRSystem {
 
     /**
      * @function
-     * @description Called when the entity component is initialized
+     * @description (async) Called when the entity component is initialized
```
Comment (hanbollar) [inline src/core/MRSystem.js:120]: i like that you added this here in the documentation

File: src/core/MREntity.js [hunk 5]
```diff
@@ -307,13 +307,13 @@ export class MREntity extends MRElement {
      * @function
      * @description Callback function of MREntity - does nothing. Is called by the connectedCallback.
      */
-    connected() {}
+    async connected() {}
 
     /**
      * @function
      * @description The connectedCallback function that runs whenever this entity component becomes connected to something else.
```
Comment (hanbollar) [inline src/core/MREntity.js:314]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/MREntity.js [hunk 6]
```diff
@@ -307,13 +307,13 @@ export class MREntity extends MRElement {
      * @function
      * @description Callback function of MREntity - does nothing. Is called by the connectedCallback.
```
Comment (hanbollar) [inline src/core/MREntity.js:308]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/MRSystem.js [hunk 7]
```diff
@@ -145,9 +142,9 @@ export class MRSystem {
      * @description Handles the component and registry aspect of the event when an entity component attaches to this system.
```
Comment (hanbollar) [inline src/core/MRSystem.js:142]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/componentSystems/AnimationSystem.js [hunk 8]
```diff
@@ -42,8 +42,18 @@ export class AnimationSystem extends MRSystem {
      * @description Called when the entity component is initialized
```
Comment (hanbollar) [inline src/core/componentSystems/AnimationSystem.js:42]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/componentSystems/AnimationSystem.js [hunk 9]
```diff
@@ -42,8 +42,18 @@ export class AnimationSystem extends MRSystem {
...
+    async attachedComponent(entity) {
         let comp = entity.components.get('animation');
+
+        await new Promise((resolve) => {
+            const interval = setInterval(() => {
+                if (entity.loaded) {
+                    clearInterval(interval);
+                    resolve();
+                }
+            }, 100); // Checks every 100ms
```
Comment (hanbollar) [inline src/core/componentSystems/AnimationSystem.js:54]: add expln in docs for why we need to keep checking

File: src/core/entities/MRDivEntity.js [hunk 10]
```diff
@@ -119,7 +119,7 @@ export class MRDivEntity extends MREntity {
      * @function
      * @description Callback function of MREntity - connects the background geometry of this item to an actual UIPlane geometry.
```
Comment (hanbollar) [inline src/core/entities/MRDivEntity.js:120]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRHyperlinkEntity.js [hunk 11]
```diff
@@ -53,7 +53,7 @@ export default class MRHyperlinkEntity extends MRTextEntity {
      * @description Callback function of MREntity - makes sure the link object is created and sets up event
```
Comment (hanbollar) [inline src/core/entities/MRHyperlinkEntity.js:53]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRImageEntity.js [hunk 12]
```diff
@@ -43,7 +43,8 @@ export class MRImageEntity extends MRMediaEntity {
      * @function
      * @description Callback function of MREntity - handles setting up this Image and associated 3D geometry style (from css) once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRImageEntity.js:44]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRLightEntity.js [hunk 13]
```diff
@@ -22,7 +22,7 @@ export class MRLightEntity extends MREntity {
      * @function
      * @description Callback function of MREntity - handles setting up this Light once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRLightEntity.js:23]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRMediaEntity.js [hunk 14]
```diff
@@ -117,7 +117,7 @@ export class MRMediaEntity extends MRDivEntity {
      * @function
      * @description Callback function of MREntity - handles setting up this media and associated 3D geometry style (from css) once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRMediaEntity.js:118]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRPanelEntity.js [hunk 15]
```diff
@@ -54,7 +54,8 @@ export class MRPanelEntity extends MRDivEntity {
      * @description Callback function of MREntity - handles setting up this Panel once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRPanelEntity.js:54]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRSkyBoxEntity.js [hunk 16]
```diff
@@ -57,7 +57,8 @@ export class MRSkyBoxEntity extends MREntity {
      * @description Lifecycle method that is called when the entity is connected.
```
Comment (hanbollar) [inline src/core/entities/MRSkyBoxEntity.js:57]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRTextAreaEntity.js [hunk 17]
```diff
@@ -78,8 +78,8 @@ export class MRTextAreaEntity extends MRTextInputEntity {
     /**
      * Overrides the connected method to include setup for handling multiline text.
```
Comment (hanbollar) [inline src/core/entities/MRTextAreaEntity.js:79]: this function needs proper jsdoc descriptions not just function body

@function
@description

and - can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRTextInputEntity.js [hunk 18]
```diff
@@ -51,7 +51,8 @@ export class MRTextInputEntity extends MRTextEntity {
      * @function
      * @description Callback function of MREntity - handles setting up this textarea once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRTextInputEntity.js:52]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRVideoEntity.js [hunk 19]
```diff
@@ -59,7 +59,8 @@ export class MRVideoEntity extends MRMediaEntity {
      * @function
      * @description Callback function of MREntity - handles setting up this video and associated 3D geometry style (from css) once it is connected to run as an entity component.
```
Comment (hanbollar) [inline src/core/entities/MRVideoEntity.js:60]: can you add `(async)` to the description like you did for the other newly async functions?

File: src/core/entities/MRTextEntity.js [hunk 20]
```diff
@@ -42,9 +42,14 @@ export class MRTextEntity extends MRDivEntity {
...
-    connected() {
-        const text = this.textContent.trim();
+    async connected() {
+        await super.connected();
+        const text = this.textContent
+            .replace(/(\n)\s+/g, '$1')
+            .replace(/(\r\n|\n|\r)/gm, ' ')
+            .trim();
         this.textObj.text = text.length > 0 ? text : ' ';
+        this.dispatchEvent(new CustomEvent('new-entity', { bubbles: true }));
```
Comment (hanbollar) [inline src/core/entities/MRTextEntity.js:52]: reminder - pls remove this `dispatchEvent` line - you told me to say this 👀

**Review by hanbollar — APPROVED** [review]
🥳 - this is definitely a chonky one - excited for it to get in

main notes:
- a few spots where `(async)` needs to be added to description
- added a few small notes on documentation to be added
- ignored the commented code and todos as those will happen in future system rewrites
- leaving this more as a cursory review since we're pushing this through for laurent's demo project

everything else `lgtm`

File: src/core/MRApp.js (hunk 1 above)
Comment (michaelthatsit) [reply src/core/MRApp.js:36]: Ah when I said order I meant of the Systems, not the events.

Comment (hanbollar) [reply src/core/MRApp.js:38]: oh lol - all good

**Pull request merged**